File: apps/android-native/app/src/main/java/com/wxxtae/pws/nativepreview/domain/TesterAuth.kt
package com.wxxtae.pws.nativepreview.domain

const val TesterAuthConfigError: String = "테스터 로그인 설정이 비어 있습니다. 관리자에게 문의해주세요."

data class TesterAuthConfig(
    val password: String,
    val allowAutoSignup: Boolean = false,
)

enum class DevTesterMode { FigmaParity, OnboardingQa, SimpleLogin }

object TesterAuth {
    const val LocalTesterSessionPrefix = "dev-"
    const val FigmaParityTesterId = "pws_dev"
    const val OnboardingQaTesterId = "pws_onboard"

    fun resolveTesterAuthConfig(rawPassword: String?): TesterAuthConfig {
        val password = rawPassword?.trim().orEmpty()
        if (password.isBlank()) throw IllegalStateException(TesterAuthConfigError)
        return TesterAuthConfig(password = password, allowAutoSignup = false)
    }

    fun resolveOptionalTesterAuthConfig(rawPassword: String?): TesterAuthConfig? {
        val password = rawPassword?.trim().orEmpty()
        if (password.isBlank()) return null
        return TesterAuthConfig(password = password, allowAutoSignup = false)
    }

    fun normalizeTesterId(testerId: String): String = testerId.trim().lowercase()

    fun testerSessionId(testerId: String): String = LocalTesterSessionPrefix + normalizeTesterId(testerId)

    fun testerEmail(testerId: String): String = "${normalizeTesterId(testerId)}@test.pws"

    fun isLocalTesterSessionId(userId: String?): Boolean = userId?.startsWith(LocalTesterSessionPrefix) == true

    fun isFigmaParitySessionId(userId: String?): Boolean = userId == testerSessionId(FigmaParityTesterId)

    fun resolveDevTesterMode(testerId: String): DevTesterMode? {
        val normalized = normalizeTesterId(testerId)
        if (normalized.isBlank()) return null
        if (normalized == FigmaParityTesterId) return DevTesterMode.FigmaParity
        if (normalized == OnboardingQaTesterId) return DevTesterMode.OnboardingQa
        return DevTesterMode.SimpleLogin
    }
}

File: shared/domain/testerAuth.ts
export const TESTER_AUTH_CONFIG_ERROR =
  '테스터 로그인 설정이 비어 있습니다. 관리자에게 문의해주세요.';

export interface TesterAuthConfig {
  password: string;
  allowAutoSignup: boolean;
}

export type DevTesterMode = 'figma-parity' | 'onboarding-qa' | 'simple-login' | null;

const LOCAL_TESTER_SESSION_PREFIX = 'dev-';
const FIGMA_PARITY_TESTER_ID = 'pws_dev';
const ONBOARDING_QA_TESTER_ID = 'pws_onboard';

export function resolveTesterAuthConfig(rawPassword: string | undefined): TesterAuthConfig {
  const password = rawPassword?.trim();

  if (!password) {
    throw new Error(TESTER_AUTH_CONFIG_ERROR);
  }

  return {
    password,
    allowAutoSignup: false,
  };
}

export function resolveOptionalTesterAuthConfig(rawPassword: string | undefined): TesterAuthConfig | null {
  const password = rawPassword?.trim();
  if (!password) return null;

  return {
    password,
    allowAutoSignup: false,
  };
}

export function normalizeTesterId(testerId: string): string {
  return testerId.trim().toLowerCase();
}

export function testerSessionId(testerId: string): string {
  return `${LOCAL_TESTER_SESSION_PREFIX}${normalizeTesterId(testerId)}`;
}

export function isLocalTesterSessionId(userId: string | undefined): boolean {
  return !!userId && userId.startsWith(LOCAL_TESTER_SESSION_PREFIX);
}

export function isFigmaParitySessionId(userId: string | undefined): boolean {
  return userId === testerSessionId(FIGMA_PARITY_TESTER_ID);
}

export function resolveDevTesterMode(testerId: string): DevTesterMode {
  const normalized = normalizeTesterId(testerId);
  if (!normalized) return null;
  if (normalized === FIGMA_PARITY_TESTER_ID) return 'figma-parity';
  if (normalized === ONBOARDING_QA_TESTER_ID) return 'onboarding-qa';
  return 'simple-login';
}
